aten/src/ATen/cuda/detail/CUDAHooks.cpp

```diff
@@ -53,9 +53,65 @@ std::unique_ptr<THCState, void (*)(THCState*)> CUDAHooks::initCUDA() const {
       });
 }
 
-std::unique_ptr<Generator> CUDAHooks::initCUDAGenerator(
-    Context* context) const {
-  return std::unique_ptr<Generator>(new CUDAGenerator(context));
+// Ensures we only call cudaGetDeviceCount only once.
+static std::once_flag num_gpu_init_flag;
+
+// Total number of gpus in the system.
+static int64_t num_gpus;
```

Comment:
Hmmmm... if we're going to cache the number of GPUs, should `CUDAHooks::getNumGPUs` use it too?

Reply:
Actually, it kind of looks like you aren't using this at all.

Reply:
Yes, we should cache the num gpus in CUDAHooks::getNumGPUs. Let's do that in the future when cleaning up `THCState`.